mt-kahypar/partition/refinement/fm/strategies/unconstrained_strategy.h

```diff
@@ -0,0 +1,503 @@
+                                                      PartitionID from,
+                                                      std::array<PartitionID, 3> parts) const {
+
+    const HypernodeWeight wu = phg.nodeWeight(u);
+    const HypernodeWeight from_weight = phg.partWeight(from);
+    PartitionID to = kInvalidPartition;
+    HyperedgeWeight to_benefit = std::numeric_limits<HyperedgeWeight>::min();
+    HypernodeWeight best_to_weight = from_weight - wu;
+    for (PartitionID i : parts) {
+      if (i != from && i != kInvalidPartition) {
+        const HypernodeWeight to_weight = phg.partWeight(i);
+        HyperedgeWeight benefit = gain_cache.benefitTerm(u, i);
+        if (upperBound >= 1 && to_weight + wu > upperBound * context.partition.max_part_weights[i]) {
+          continue;
+        } else if (to_weight + wu > context.partition.max_part_weights[i] && penaltyFactor > 0) {
+          const Gain imbalance_penalty = estimatePenalty(i, to_weight, wu);
+          if (imbalance_penalty == std::numeric_limits<Gain>::max()) {
+            continue;
+          }
+          benefit -= std::ceil(penaltyFactor * imbalance_penalty);
+        }
+        if ( benefit > to_benefit || ( benefit == to_benefit && to_weight < best_to_weight ) ) {
+          to_benefit = benefit;
+          to = i;
+          best_to_weight = to_weight;
+        }
+      }
+    }
+    ASSERT(from == phg.partID(u));
+    const Gain gain = to != kInvalidPartition ? to_benefit - gain_cache.penaltyTerm(u, from)
+                                              : std::numeric_limits<HyperedgeWeight>::min();
+    return std::make_pair(to, gain);
+  }
+
+  MT_KAHYPAR_ATTRIBUTE_ALWAYS_INLINE
+  Gain estimatePenalty(PartitionID to, HypernodeWeight to_weight, HypernodeWeight wu) const {
+    HypernodeWeight virtual_delta = sharedData.unconstrained.virtual_weight_delta[to].load(std::memory_order_relaxed)
+                                    + localVirtualWeightDelta[to];
+    HypernodeWeight initial_imbalance = to_weight + virtual_delta - context.partition.max_part_weights[to];
+    return sharedData.unconstrained.estimatePenaltyForImbalancedMove(to, initial_imbalance, wu);
+  }
+
+  const Context& context;
+
+  FMStats& runStats;
+
+  FMSharedData& sharedData;
+
+  // ! Priority Queue that contains for each block of the partition
+  // ! the vertex with the best gain value
+  BlockPriorityQueue& blockPQ;
+
+  // ! From PQs -> For each block it contains the vertices (contained
+  // ! in that block) touched by the current local search associated
+  // ! with their gain values
+  vec<VertexPriorityQueue>& vertexPQs;
+
+  // ! Virtual block weights are saved as delta to the actual block weight. They
+  // ! are necessary to ensure a reasonable penalty estimation in some edge cases.
+  VirtualWeightMap localVirtualWeightDelta;
+
+  double penaltyFactor;
+  double upperBound;
+};
+
+
+template<typename TypeTraits, typename GainTypes>
+class UnconstrainedStrategy: public IFMStrategy {
```

Comment:
would split this into two separate files "unconstrained_fm_strategy.h" and "local_unconstrained_fm_strategy.h"